fix(dogfood): protocol review changes

MaxMustermann2 opened the pull request:
commit f3ac6cd53574efbeb60d2a53fc93abd8ddde2afd (HEAD -> protocol-review/dogfood, origin/protocol-review/dogfood)
Author: MaxMustermann2 <82761650+MaxMustermann2@users.noreply.github.com>
Date:   Sun Dec 14 14:51:24 2025 +0000

    fix(dogfood): schedule completion of key changes
    
    Previously, we incorrectly relied on `GetImuachainValidator` to
    determine whether a consensus key event can be completed immediately or
    scheduled for a later epoch.
    
    If a validator is currently jailed, the validator object in x/dogfood
    cannot be found; thus, the consensus key event would be processed
    immediately. However, doing so would mean that any slashable actions
    that took place prior to the jailing event would escape slashing. To
    avoid that, the scheduling is always applied.

commit a5038315d48e09e62fc69d4566d084fa2041eee2
Author: MaxMustermann2 <82761650+MaxMustermann2@users.noreply.github.com>
Date:   Sun Dec 14 13:38:36 2025 +0000

    fix(dogfood): remove undelegation tracking
    
    The undelegation state is now handled by x/delegation entirely so we
    remove it from x/dogfood entirely. This is a breaking change since it
    alters the storage layout, however, functionality was disabled much
    earlier


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **Refactor**
  * Undelegation maturation removed from the module and genesis; related query RPCs and CLI subcommands deleted.
  * Operator key lifecycle now always schedules opt-out/pruning and defers retries to the next epoch.
  * Delegation handling routed through the delegation keeper (hook composition simplified).

* **Breaking Changes**
  * Genesis API/schema changed: undelegation maturities removed from the public genesis/state and constructor signature.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MaxMustermann2 committed:
fix(dogfood): remove undelegation tracking

The undelegation state is now handled by x/delegation entirely so we
remove it from x/dogfood entirely. This is a breaking change since it
alters the storage layout, however, functionality was disabled much
earlier

MaxMustermann2 committed:
fix(dogfood): schedule completion of key changes

Previously, we incorrectly relied on `GetImuachainValidator` to
determine whether a consensus key event can be completed immediately or
scheduled for a later epoch.

If a validator is currently jailed, the validator object in x/dogfood
cannot be found; thus, the consensus key event would be processed
immediately. However, doing so would mean that any slashable actions
that took place prior to the jailing event would escape slashing. To
avoid that, the scheduling is always applied.

MaxMustermann2 committed:
fix(lint)

MaxMustermann2 committed:
fix(swagger)

MaxMustermann2 committed:
fix(dogfood): use correct context

bwhour commented on x/dogfood/keeper/impl_delegation_hooks.go:
```diff
@@ -1,47 +1 @@
 package keeper
-
```
it seems there is no implementation logic in this file

MaxMustermann2 committed:
Merge upstream/develop

MaxMustermann2 committed:
fix(dogfood): build and comments

bwhour commented on x/dogfood/keeper/opt_out.go:
x/dogfood/types/keys.go defines these prefixes:

- The prefix 0x02 stores: "Which operators completed opt-out in which epoch" (type is AccountAddresses, i.e., operator account addresses)
- The prefix 0x04 stores: "Which consensus addresses need to be cleaned up in which epoch" (type is ConsensusAddresses, i.e., consensus addresses)

Here **GetAllConsAddrsToPrune —— should read consensus address cleanup data**

bwhour commented on x/dogfood/keeper/abci.go:
```diff
@@ -40,17 +40,7 @@ func (k Keeper) EndBlock(ctx sdk.Context) []abci.ValidatorUpdate {
+	// let the operator module know that the opt out has finished.
 	optOuts := k.GetPendingOptOuts(ctx)
 	for _, addr := range optOuts.GetList() {
 		err := k.operatorKeeper.CompleteOperatorKeyRemovalForChainID(
```
If _**CompleteOperatorKeyRemovalForChainID**_ returns an error (such as ErrOperatorNotRemovingKey), the operator's opt-out will be permanently lost and cannot be retried. This may lead to inconsistent key status of the operator — which is a security risk.

MaxMustermann2 replied:
Rescheduled this for the next epoch upon error.

cloud8little committed:
Merge branch 'develop' into protocol-review/dogfood

MaxMustermann2 committed:
fix: respond to review comments

MaxMustermann2 commented on x/dogfood/keeper/impl_operator_hooks.go:
```diff
@@ -49,82 +49,79 @@ func (h OperatorHooksWrapper) AfterOperatorKeyReplaced(
 		if avs == dogfoodAVSAddr {
-			found, wrappedKey, err := h.keeper.operatorKeeper.GetOperatorConsKeyForChainID(ctx, operator, chainIDWithoutRevision)
+			// check whether a consensus key exists for the operator
+			found, wrappedKey, err := h.keeper.operatorKeeper.GetOperatorConsKeyForChainID(
+				ctx, operator, chainIDWithoutRevision,
+			)
+			// ideally, this will not happen - except if the operator is opting out?
 			if !found || err != nil {
-				ctx.Logger().Error("AfterSlash the consensus key isn't found by the chainIDWithoutRevision and operator address", "operatorAddr", operator, "chainIDWithoutRevision", chainIDWithoutRevision, "err", err)
+				h.keeper.Logger(ctx).Error(
+					"could not find consensus key for operator",
+					"operatorAddr", operator,
+					"chainIDWithoutRevision", chainIDWithoutRevision,
+					"err", err,
+				)
 				return
```
Corrected the comment.

MaxMustermann2 committed:
fix(operator): check prev key existence

MaxMustermann2 commented on x/dogfood/keeper/abci.go:
```diff
@@ -36,28 +36,27 @@
 		if err != nil {
 			logger.Error("error completing operator key removal", "error", err)
+			// reschedule it for the next epoch
+			k.AppendOptOutToFinish(ctx, nextEpochNumber, addr)
```
This panic is acceptable; en error in the marshal/unmarshal process indicates hopeless corruption of the underlying database.

MaxMustermann2 commented:
@coderabbitai review

MaxMustermann2 committed:
fix: respond to AI review

1. Collect failed operator key removals into an array
2. Check for epoch presence
3. Do not silently complete operator key removal if no key found

MaxMustermann2 commented on x/dogfood/keeper/abci.go:
```diff
@@ -36,30 +38,54 @@
+	// epoch identifier in the MsgServer. thus, it is guaranteed to exist.
+	epochInfo, found := k.epochsKeeper.GetEpochInfo(ctx, epochID)
+	if !found {
+		panic(fmt.Sprintf("epoch %s not found", epochID))
```
See comment above the function, acceptable risk.

bwhour reviewed (approved):
LGTM.
